activesupport/lib/active_support/notifications/event_assertions.rb

```diff
@@ -0,0 +1,79 @@
+# typed: true
+# frozen_string_literal: true
+
+module ActiveSupport
+  module Notifications
+    module EventAssertions
+      # Assert an event was emitted with the given +name+ and +payload+.
+      #
+      # You can assert that an event was emitted by passing a name, an optional
+      # payload, and a block. While the block is executed, if a matching event
+      # is emitted, the assertion will pass. Otherwise, it will fail.
+      #
+      #     assert_event("post.submitted", title: "Cool Post") do
+      #       post.submit(title: "Cool Post") # => emits matching ActiveSupport::Notifications::Event
+      #     end
+      #
+      def assert_event(name, payload = nil, &block)
```

Comment:
`event` is a fairly generic term. Since this is an Active Support test assertion, it's possible that it could collide with application-side domain terminology.

Have you considered other terminology from the domain of the [ActiveSupport::Notifications](https://edgeapi.rubyonrails.org/classes/ActiveSupport/Notifications.html) module that's more specific? Instances of `Event` are made available to the blocks, but those events collect instrumentation data that's being published to subscribers.

I'm struggling to suggest a specific improvement, but I wonder if there are alternatives that are worthwhile to consider. Maybe even `assert_notification_event` or `assert_instrumentation_event` could work.

Reply:
Have you explored existing patterns of use throughout the rest of the `rails/rails` test suites? There are various calls throughout that could utilize the assertions you're introducing in this PR. After a cursory search, I found:

* https://github.com/rails/rails/blob/99a6365c30a745f7647ccdbf4ba084530dca63eb/actioncable/test/channel/base_test.rb#L201-L263
* https://github.com/rails/rails/blob/99a6365c30a745f7647ccdbf4ba084530dca63eb/actioncable/test/server/broadcasting_test.rb#L15-L53
* https://github.com/rails/rails/blob/99a6365c30a745f7647ccdbf4ba084530dca63eb/actionmailbox/test/unit/mailbox/notifications_test.rb#L9-L31
* https://github.com/rails/rails/blob/99a6365c30a745f7647ccdbf4ba084530dca63eb/actionmailer/test/base_test.rb#L966-L992

Some tests even utilize the [ActiveSupport::Notifications.subscribed](https://edgeapi.rubyonrails.org/classes/ActiveSupport/Notifications.html#method-c-subscribed) variety of call.

* https://github.com/rails/rails/blob/99a6365c30a745f7647ccdbf4ba084530dca63eb/actionpack/test/dispatch/routing/instrumentation_test.rb#L27-L32
* https://github.com/rails/rails/blob/99a6365c30a745f7647ccdbf4ba084530dca63eb/actionpack/test/dispatch/middleware_stack_test.rb#L195-L213

For what it's worth, my anecdotal findings make me think that the favored terminology for the concept is [instrumentation](https://github.com/search?q=repo%3Arails%2Frails%20instrumentation_test.rb&type=code), so it could be worthwhile to incorporate that into the name of the assertion methods or module.

Reply:
> `event` is a fairly generic term. Since this is an Active Support test assertion, it's possible that it could collide with application-side domain terminology.

Completely agree. Been on-and-off thinking about what a more specific name might look like. Considering the Rails API docs states the following:

> `ActiveSupport::Notifications` provides an instrumentation API for Ruby.

I do agree that `assert_instrumentation_event` seems pretty accurate and would have a pretty low chance of name collision as well.

> Have you explored existing patterns of use throughout the rest of the `rails/rails` test suites? There are various calls throughout that could utilize the assertions you're introducing in this PR.

That cleanup opportunity was mentioned when I first proposed upstreaming this module. It's definitely something I'd like to do, a sweep over the codebase. Though I will probably do so in a follow-up PR. May look into opening one soon, stacked on this, as the implementation in here is slowly but surely maturing.

Much appreciate the cursory search. :pray:

Reply:
Briefly looked into updating the existing `ActiveSupport::Notifications` usage in tests. It looks like we should probably add an `unsubscribe` in these helpers as that's done in said tests. Should ideally ensure no side effects from using these helpers as well.

Have pushed a change to add `unsubscribe` the shared `capture_events` method.